doc/write-a-wasm-extension-with-cpp.md

```diff
@@ -0,0 +1,178 @@
+http_archive(
+    name = "proxy_wasm_cpp_sdk",
+    strip_prefix = "proxy-wasm-cpp-sdk-2ccab8dfc0b435f3d7641cef7683c7861e23d179",
+    url = "https://github.com/proxy-wasm/proxy-wasm-cpp-sdk/archive/2ccab8dfc0b435f3d7641cef7683c7861e23d179.tar.gz",
+)
+
+load("@proxy_wasm_cpp_sdk//bazel/dep:deps.bzl", "wasm_dependencies")
+
+wasm_dependencies()
+
+load("@proxy_wasm_cpp_sdk//bazel/dep:deps_extra.bzl", "wasm_dependencies_extra")
+
+wasm_dependencies_extra()
+```
+
+You can rename the workspace to anything relevent to your extension. Other dependencies could also be imported as needed, such as JSON library, base64 library, etc. See the top level `WORKSPACE` file for examples.
+
+## Step 2: Set up extension scaffold
+
+Copy the [scaffold plugin code](../extensions/scaffold), which includes a plugin header file `plugin.h`, a source file `plugin.cc`, and a Bazel BUILD file `BUILD`. Give the Wasm binary a meaninful name. In the example folder, we rename it to `example.wasm`.
+
+The plugin source code implements the follows:
+* A root context, which has the same lifetime as the WebAssembly VM, handles plugin configuration, and acts as the context for any interactions with Envoy host which outlives individual streams, e.g. timer tick, HTTP or gRPC call out.
+* A stream context, which has the same lifetime as each request stream and acts as the target for callbacks from each stream.
+
+After these files are copied, try build your Wasm extension with:
+```
+bazel build //:example.wasm
+```
+
+After the build finishes, under the generated `bazel-bin/` folder, you should be able to find a Wasm file named as `example.wasm`. Congratulations! you just successfully built an Envoy Wasm extension!
+
+## Step 3: Fill in extension logic
+
+The current Wasm extension is a no-op extension, callbacks need to be filled in order to make it actually function. (TODO: add doc into proxy wasm cpp sdk about available callbacks). In this example plugin, `OnResponseHeaders` is implemented to inject a response header. Specifically, followings are added into header and source file:
+
+`plugin.h`
+```cpp
+class PluginContext : public Context {
+public:
+ ...
+ FilterHeadersStatus onResponseHeaders(uint32_t, bool) override;
+ ...
+};
+```
+
+`plugin.cc`
+```cpp
+FilterHeadersStatus PluginContext::onResponseHeaders(uint32_t, bool) {
+ addResponseHeader("X-Wasm-custom", "foo");
+ return FilterHeadersStatus::Continue;
+}
+```
+
+After adding the method, build the extension again with the bazel command mentioned in step 2 and the `example.wasm` file should be updated now with the new logic.
+
+## Step 4: Write integration test
+
+To enable quick iteration and deploy your extension with confidence, it is **highly** recommended to write integration tests with your extension and the same Envoy binary Istio proxy runs. To achieve this, the same Golang integration test framework for Telemetry v2 filter could be used, which at high level does the follows:
```

Comment:
add a link to the telemetry v2 integration framework?

Reply:
Somehow go doc does not have istio.io/proxy. Trying to figure out why is it..

Reply:
Actually it was there: https://godoc.org/github.com/istio/proxy/test/envoye2e